docs/source/high-level-arch.rst

```diff
@@ -0,0 +1,122 @@
+
+******************
+ Embedding Tables
+******************
+
+For TorchRec to figure out what to recommend, we need to be able to
+represent entities and their relationships, this is what embeddings are
+used for. Embeddings are vectors of real numbers in a high dimensional
+space used to represent meaning in complex data like words, images, or
+users. An embedding table is an aggregation of multiple embeddings into
+one matrix. Most commonly, embedding tables are represented as a 2D
+matrix with dimensions (B, N).
+
+-  *B* is the number of embeddings stored by the table
+-  *N* is number of dimensions per embedding.
+
+Each of *B* can also be referred to as an ID (representing information
+such as movie title, user, ad, and so on), when accessing an ID we are
+returned the corresponding embedding vector which has size of embedding
+dimension *N*.
+
+There is also the choice of pooling embeddings, often, we’re looking up
+multiple rows for a given feature which gives rise to the question of
+what we do with looking up multiple embedding vectors. Pooling is a
+common technique where we combine the embedding vectors, usually through
+sum or mean of the rows, to produce one embedding vector. This is the
+main difference between the PyTorch ``nn.Embedding`` and
+``nn.EmbeddingBag``.
+
+PyTorch represents embeddings through ``nn.Embedding`` and
+``nn.EmbeddingBag``. Building on these modules, TorchRec introduces
+``EmbeddingCollection`` and ``EmbeddingBagCollection``, which are
+collections of the corresponding PyTorch modules. This extension enables
+TorchRec to batch tables and perform lookups on multiple embeddings in a
+single kernel call, improving efficiency.
+
+Here is the end-to-end flow diagram that describes how embeddings are
+used in the training process for recommendation models:
+
+.. figure:: _static/img/full_training_loop.png
+   :alt: Demonstrating the full training loop from embedding lookup to optimizer update in backward
+   :align: center
+
+   *Figure 2. TorchRec End-to-end Embedding Flow*
+
+In the diagram above, we show the general TorchRec end to end embedding
+lookup process. In the forward we do the embedding lookup and pooling,
+in the backward we compute the gradients of the output embedding lookups
+and apply them to the appropriate embedding table gradients and pass it
+into the optimizer. Note here, these gradients are grayed out since we
+don’t fully materialize these into memory and instead fuse them with the
+optimizer update. This results in a significant memory reduction which
+we detail later in the optimizer section.
+
+We recommend going through the TorchRec Concepts page as well to get a
+understanding of the fundamentals of how everything ties together
+end-to-end. It contains lots of useful information to get the most out
+of TorchRec.
+
+**********
+ See also
+**********
+
+-  `What is Distributed Data Parallel (DDP) Tutorial
+   <https://pytorch.org/tutorials/beginner/ddp_series_theory.html>`_
```

Comment:
Maybe also add a link to the future Concepts section?

Reply:
yup was planning to add this in the concepts doc PR